norse/torch/module/receptive_field.py

```diff
@@ -4,85 +4,260 @@
-                weights.append(in_weights)
-            weights = torch.concat(weights, 1).permute(1, 0, 2, 3)
-
-        self.conv = torch.nn.Conv2d(in_channels, self.out_channels, size, **kwargs)
-        self.conv.weight = torch.nn.Parameter(weights)
+        self.rf_parameters_previous = torch.zeros_like(self.rf_parameters)
+
+        self.in_channels = in_channels
+        self.size = size
+        self.aggregate = aggregate
+        self.domain = domain
+        self.out_channels = (
+            rf_parameters.shape[0]
+            if aggregate
+            else rf_parameters.shape[0] * in_channels
+        )
+
+        # Register update and update the fields for the first time
+        self.has_updated = True
+        self._update_weights()
+
+        if optimize_fields:
+
+            def update_hook(m, gi, go):
+                self.has_updated = True
+
+            self.register_full_backward_hook(update_hook)
+
+    def _update_weights(self):
+        if self.has_updated:
+            if not torch.all(torch.eq(self.rf_parameters_previous, self.rf_parameters)):
+                # Reset the flag
+                self.has_updated = False
+                self.rf_parameters_previous = self.rf_parameters.detach().clone()
+                # Calculate new weights
+                fields = spatial_receptive_fields_with_derivatives(
+                    self.rf_parameters,
+                    size=self.size,
+                    domain=self.domain,
+                )
+                if self.aggregate:
+                    self.out_channels = fields.shape[0]
+                    self.weights = fields.unsqueeze(1).repeat(1, self.in_channels, 1, 1)
+                else:
+                    self.out_channels = self.fields.shape[0] * self.in_channels
+                    empty_weights = torch.zeros(
+                        self.in_channels,
+                        fields.shape[0],
+                        self.size,
+                        self.size,
+                        device=self.rf_parameters.device,
+                    )
+                    weights = []
+                    for i in range(self.in_channels):
+                        in_weights = empty_weights.clone()
+                        in_weights[i] = fields
+                        weights.append(in_weights)
+                    self.weights = torch.concat(weights, 1).permute(1, 0, 2, 3)
+                self.weights.requires_grad_(True)
 
     def forward(self, x: torch.Tensor):
-        return self.conv(x)
+        self._update_weights()  # Update weights if necessary
+        return torch.nn.functional.conv2d(x, self.weights, bias=None)
+
+
+class SampledSpatialReceptiveField2d(torch.nn.Module):
+    """
+    Creates a spatial receptive field as 2-dimensional convolutions, sampled over a set of scales,
+    angles, ratios, and derivatives.
+    This module allows for the optimization of the scales, angles, and ratios (not derivatives)
```

Comment:
Maybe it should be specified that the optimization is happening at the input parameters and not at individual kernels
For example:
     This module allows for the optimization of the input parameters of scales, angles, and ratios (not derivatives)